site/content/blog/2020-02-10-maintaining-consistency-in-codebases-with-go-vet.md

```diff
@@ -0,0 +1,181 @@
+---
+title: Maintaining consistency in codebases with Go Vet
+slug: maintaining-consistency-in-codebases-with-go-vet
+date: 2020-02-10
+categories:
+    - "go"
+author: Jesús Espino
+github: jespino
+community: jesus.espino
+---
+
+At Mattermost we are a bit obsessed with the quality of our code, and one of
+the things that we love is consistent and idiomatic Go. In that sense, go vet
+is a great tool, because it checks a lot of things for us (useless assignments,
+wrong number of Printf parameters, correct format in struct tags, etcetera) as
+part of our CI pipeline, helping us and our contributors to write better code.
+
+The main restriction of go vet is that there are a limited number of absolute
+truths about what is right and what is wrong and so the go vet checks are very
+general.
+
+In Mattermost we have a way of doing certain things like logging, test
+assertions or license headers. We love to keep code consistent so we work hard
+to avoid reintroducing old patterns in the code.
+
+I think the best way to explain it is with an example.
+
+Some time ago we decided to migrate our logging approach to use structured
+logging everywhere, but the migration wasn't complete and the old pattern
+(logging without structured approach) was starting to appear again, in some
+cases mimicking code that wasn’t migrated to the new pattern yet, and in other
+cases by coincidence because the old pattern was one of the obvious approaches
+that you can follow to present data in the logs.
+
+So... go vet to the rescue! Go vet comes with a good bunch of checks already,
+but you can extend it by adding your own specific checks.
+
+We created [our own check](https://github.com/mattermost/mattermost-govet/blob/master/structuredLogging/structuredLogging.go)
+of go vet avoiding any string building with `fmt.Sprintf` calls as part of the
+calls to our log library. With that check implemented we were able to detect
+all the cases in the code where we were doing not-structured logging and
+replace them with the proper structured logging approach. After that, we added
+that check to our CI pipeline to ensure that the pattern is not reintroduced
+accidentally by us or by any contributor.
+
+Another interesting example is our approach to improve the quality of the test
```

Comment:
```suggestion
Another interesting example is our approach to improve the consistency of the test
```

Reply:
Actually, is more about the quality of the test assertions, I was thinking about how "semantic are they" but at the same time we were improving consistency too.